File: android/app/src/main/java/com/noop/protocol/DataRange.kt
package com.noop.protocol

/**
 * GET_DATA_RANGE frame parsing. Byte-identical twin of Swift `WhoopProtocol.DataRange` — the newest-record
 * read gates automatic sync (`isFutureDatedNewest` -> BackfillPolicy), so its value must match across
 * platforms. Extracted from WhoopBleClient (#286 follow-up) so both a Kotlin JVM test and a WhoopProtocol
 * `swift test` pin the parity.
 */
object DataRange {
    /**
     * The newest plausible unix time banked by the strap, from a GET_DATA_RANGE frame. Scans EVERY byte
     * offset (the newest-record u32 isn't on a fixed grid — it sits at byte offset 8 on WHOOP 4, off the
     * old aligned-from-7 scan), keeps the newest word in a plausible unix window (2023-11..2030-03),
     * preferring the newest that is NOT implausibly future (> wallNowUnix + futureSkewSeconds) so a garbage
     * future word can't latch and stall auto-sync (#451/#928/#1012). Falls back to the newest-any word so a
     * genuinely future-dated RTC is still surfaced downstream. null only for a too-short frame or no
     * plausible word. Mirrors Swift `DataRange.newestUnix`.
     */
    fun newestUnix(frame: ByteArray, wallNowUnix: Long, futureSkewSeconds: Long): Long? {
        if (frame.size < 4) return null
        val futureCutoff = wallNowUnix + futureSkewSeconds
        var newestNotFuture: Long? = null
        var newestAny: Long? = null
        var i = 0
        while (i + 4 <= frame.size) {
            val w = (frame[i].toLong() and 0xFFL) or
                ((frame[i + 1].toLong() and 0xFFL) shl 8) or
                ((frame[i + 2].toLong() and 0xFFL) shl 16) or
                ((frame[i + 3].toLong() and 0xFFL) shl 24)
            if (w in 1_700_000_000L..1_900_000_000L) {
                newestAny = maxOf(newestAny ?: 0L, w)
                if (w <= futureCutoff) newestNotFuture = maxOf(newestNotFuture ?: 0L, w)
            }
            i += 1
        }
        return newestNotFuture ?: newestAny
    }
}

File: Packages/WhoopProtocol/Sources/WhoopProtocol/DataRange.swift
import Foundation

/// GET_DATA_RANGE frame parsing. Pure + byte-identical to the Kotlin twin `com.noop.protocol.DataRange`.
///
/// Extracted from `BLEManager`/`WhoopBleClient` (#286 follow-up): the newest-record read gates automatic
/// sync (`isFutureDatedNewest` → `BackfillPolicy`), so its value must stay byte-identical across platforms.
/// Living here means both a WhoopProtocol `swift test` and the Kotlin JVM test pin the parity — the app-target
/// copies were previously testable on the Kotlin side only.
public enum DataRange {
    /// The newest plausible unix time banked by the strap, from a GET_DATA_RANGE frame.
    ///
    /// Scans EVERY byte offset — the newest-record `u32` isn't on a fixed grid (it sits at byte offset 8 on
    /// WHOOP 4.0, off the old aligned-from-7 scan, which straddled it and returned nil). Keeps the newest
    /// word inside a plausible unix window (2023-11…2030-03), preferring the newest that is NOT implausibly
    /// future (> `wallNowUnix + futureSkewSeconds`) so a garbage future word can't latch and stall auto-sync
    /// (#451/#928/#1012). Falls back to the newest-any word so a genuinely future-dated RTC is still surfaced
    /// downstream and the future-date guard still fires. Returns nil only for a too-short frame or no
    /// plausible word at all.
    public static func newestUnix(from frame: [UInt8], wallNowUnix: Int, futureSkewSeconds: Int) -> Int? {
        guard frame.count >= 4 else { return nil }
        let futureCutoff = wallNowUnix + futureSkewSeconds
        var newestNotFuture: Int? = nil
        var newestAny: Int? = nil
        var i = 0
        while i + 4 <= frame.count {
            let w = Int(frame[i]) | Int(frame[i + 1]) << 8 | Int(frame[i + 2]) << 16 | Int(frame[i + 3]) << 24
            if w >= 1_700_000_000 && w <= 1_900_000_000 {
                newestAny = max(newestAny ?? 0, w)
                if w <= futureCutoff { newestNotFuture = max(newestNotFuture ?? 0, w) }
            }
            i += 1
        }
        return newestNotFuture ?? newestAny
    }
}
